File: Pather/Pather/Game.ts
/// <reference path="constants.ts" />
/// <reference path="Person.ts" />

module Pather {
   
    export class Game {
        nextGameTick: number;
        grid: boolean[][];
        people: Person[];
        curTick: number;

        canvas: HTMLCanvasElement;
        context: CanvasRenderingContext2D;
         


        constructor() {
            this.nextGameTick = (new Date).getTime();
            this.constructGrid();
            this.people = [];

            var sal = new Person(this);
            this.people.push(sal); 

            this.curTick = (new Date).getTime();


            this.canvas = <HTMLCanvasElement>document.getElementById("canvas");
            this.context = this.canvas.getContext("2d"); 
            this.canvas.onmousedown = (ev) => {
                var person = this.people[0];
                person.rePathFind((ev.offsetX / Constants.squareSize) | 0, (ev.offsetY / Constants.squareSize) | 0);
            };

        }

        constructGrid() {
            this.grid = [];
            for (var i = 0; i < Constants.numberOfSquares; i++) {
                var gridRow = [];
                for (var a = 0; a < Constants.numberOfSquares; a++) {
                    gridRow.push((Math.random() * 100 < 15) ? 0 : 1);
                }
                this.grid.push(gridRow);
            }
        }


        init() {
            for (var i = 0; i < this.people.length; i++) {
                this.people[i].init(0, 0);
            }

            window.setTimeout(() => this.tick(), Constants.gameTicks);
            (<any>window).requestAnimFrame(() => this.draw());

        }

        tick() {
            window.setTimeout(() => this.tick(), Constants.gameTicks);
            var v = (new Date).getTime();
            this.nextGameTick += (v - this.curTick);
            this.curTick = v;

            for (var i = 0; i < this.people.length; i++) {
                this.people[i].tick();
            }
        }


        draw() {
            (<any>window).requestAnimFrame(() => this.draw());
            this.context.save();
            this.context.fillStyle = '#000000';
            this.context.fillRect(0, 0, 1200, 1200);
            this.context.restore();

            for (var y = 0; y < Constants.numberOfSquares; y++) {
                for (var x = 0; x < Constants.numberOfSquares; x++) {
                    if (this.grid[x][y]) {
                        this.context.save();
                        this.context.lineWidth = 5;
                        this.context.strokeStyle = 'white';
                       // this.context.strokeRect(x * Constants.squareSize, y * Constants.squareSize, Constants.squareSize, Constants.squareSize);
                        this.context.restore();
                    }
                }
            }
            for (var y = 0; y < Constants.numberOfSquares; y++) {
                for (var x = 0; x < Constants.numberOfSquares; x++) {
                    if (!this.grid[x][y]) {
                        this.context.save();
                        this.context.lineWidth = 5;
                        this.context.strokeStyle = 'blue';
                        this.context.strokeRect(x * Constants.squareSize, y * Constants.squareSize, Constants.squareSize, Constants.squareSize);
                        this.context.restore();
                    }

                }
            }

            for (var y = 0; y < Constants.numberOfSquares; y++) {
                for (var x = 0; x < Constants.numberOfSquares; x++) {
                    for (var i = 0; i < this.people.length; i++) {
                        for (var j = 0; j < this.people[i].path.length; j++) {

                            if (this.people[i].path[j].x == x && this.people[i].path[j].y == y) {
                                this.context.save();
                                this.context.lineWidth = 5;
                                this.context.strokeStyle = 'pink';
//                                this.context.strokeRect(x * Constants.squareSize, y * Constants.squareSize, Constants.squareSize, Constants.squareSize);
                                this.context.restore();

                            }

                        }

                    }


                }
            }

            var interpolatedTime = (((new Date).getTime() - this.nextGameTick) / Constants.gameTicks);


            for (var i = 0; i < this.people.length; i++) {
                this.people[i].draw(this.context, interpolatedTime);
            }

        }

    }
}

File: Pather.Client/Game.cs
using System;
using System.Collections.Generic;
using System.Html;
using System.Html.Media.Graphics;

namespace Pather.Client
{
    public class Game
    {
        public long NextGameTick { get; set; }
        public int[][] Grid { get; set; }
        public List<Person> People { get; set; }
        public long CurTick { get; set; }

        public CanvasElement Canvas { get; set; }
        public CanvasRenderingContext2D Context { get; set; }

        public Game()
        {
            NextGameTick = new DateTime().GetTime();
            ConstructGrid();
            People = new List<Person>();
            var sal = new Person(this);
            People.Add(sal);

            CurTick = new DateTime().GetTime();

            Canvas = (CanvasElement)Document.GetElementById("canvas");
            Context = (CanvasRenderingContext2D)Canvas.GetContext(CanvasContextId.Render2D);

            Canvas.OnMousedown = (ev) =>
            {
                var person = People[0];
                var @event = (dynamic)ev;
                person.RePathFind(((int)@event.offsetX) / Constants.SquareSize, ((int)@event.offsetY) / Constants.SquareSize);
            };
        }

        public void ConstructGrid()
        {
            Grid = new int[Constants.NumberOfSquares][];
            for (int x = 0; x < Constants.NumberOfSquares; x++)
            {
                Grid[x] = new int[Constants.NumberOfSquares];
                for (int y = 0; y < Constants.NumberOfSquares; y++)
                {
                    Grid[x][y] = (Math.Random() * 100 < 15)?0:1;
                }
            }
        }

        public void Init()
        {
            foreach (var person in People)
            {
                person.Init(0, 0);
            }
            Window.SetTimeout(Tick, Constants.GameTicks);
            Window.RequestAnimationFrame((a) => Draw());
        }

        public void Tick()
        {
            Window.SetTimeout(Tick, Constants.GameTicks);

            var v = new DateTime().GetTime();
            NextGameTick += v - CurTick;
            CurTick = v;
            foreach (var person in People)
            {
                person.Tick();
            }
        }

        public void Draw()
        {
            Window.RequestAnimationFrame((a) => Draw());

            Context.Save();
            Context.FillStyle = "black";
            Context.FillRect(0, 0, 1200, 1200);
            Context.Restore();
            for (var y = 0; y < Constants.NumberOfSquares; y++)
            {
                for (var x = 0; x < Constants.NumberOfSquares; x++)
                {
                    if (Grid[x][y]==1)
                    {
                        Context.Save();
                        Context.LineWidth = 5;
                        Context.StrokeStyle = "white";
//                        Context.StrokeRect(x * Constants.SquareSize, y * Constants.SquareSize, Constants.SquareSize, Constants.SquareSize);
                        Context.Restore();
                    }
                }
            }
            for (var y = 0; y < Constants.NumberOfSquares; y++)
            {
                for (var x = 0; x < Constants.NumberOfSquares; x++)
                {
                    if (Grid[x][y]==0)
                    {
                        Context.Save();
                        Context.LineWidth = 5;
                        Context.StrokeStyle = "blue";
                        Context.StrokeRect(x * Constants.SquareSize, y * Constants.SquareSize, Constants.SquareSize, Constants.SquareSize);
                        Context.Restore();
                    }
                }
            }

            var interpolatedTime = (((new DateTime()).GetTime() - NextGameTick) / (double)Constants.GameTicks);


            foreach (var person in People)
            {
                person.Draw(Context, interpolatedTime);
            }
        }
    }
}